pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -314,6 +328,76 @@ func TestCreateAndDeleteOfAffinityAssistantPerWorkspaceOrDisabled(t *testing.T)
 	}
 }
 
+func TestCreateOrUpdateAffinityAssistantsAndPVCs_Failure(t *testing.T) {
+	testCases := []struct {
+		name, failureType string
+		aaBehavior        aa.AffinityAssitantBehavior
+		expectedErr       error
+	}{{
+		name:        "affinity assistant creation failed - per workspace",
+		failureType: "statefulset",
+		aaBehavior:  aa.AffinityAssistantPerWorkspace,
+		expectedErr: fmt.Errorf("%w: [failed to create StatefulSet affinity-assistant-4cf1a1c468: error creating statefulsets]", ErrAffinityAssistantCreationFailed),
+	}, {
+		name:        "affinity assistant creation failed - per pipelinerun",
+		failureType: "statefulset",
+		aaBehavior:  aa.AffinityAssistantPerPipelineRun,
+		expectedErr: fmt.Errorf("%w: [failed to create StatefulSet affinity-assistant-426b306c50: error creating statefulsets]", ErrAffinityAssistantCreationFailed),
+	}, {
+		name:        "pvc creation failed - per workspace",
+		failureType: "pvc",
+		aaBehavior:  aa.AffinityAssistantPerWorkspace,
+		expectedErr: fmt.Errorf("%w: failed to create PVC pvc-b9eea16dce: error creating persistentvolumeclaims", ErrPvcCreationFailed),
+	}, {
+		name:        "pvc creation failed - disabled",
+		failureType: "pvc",
+		aaBehavior:  aa.AffinityAssistantDisabled,
+		expectedErr: fmt.Errorf("%w: failed to create PVC pvc-b9eea16dce: error creating persistentvolumeclaims", ErrPvcCreationFailed),
+	}}
+
+	for _, tc := range testCases {
+		ctx := context.Background()
+		kubeClientSet := fakek8s.NewSimpleClientset()
+		c := Reconciler{
+			KubeClientSet: kubeClientSet,
+			pvcHandler:    volumeclaim.NewPVCHandler(kubeClientSet, zap.NewExample().Sugar()),
+		}
+
+		switch tc.failureType {
+		case "pvc":
+			c.KubeClientSet.CoreV1().(*fake.FakeCoreV1).PrependReactor("create", "persistentvolumeclaims",
+				func(action testing2.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &corev1.PersistentVolumeClaim{}, errors.New("error creating persistentvolumeclaims")
+				})
+		case "statefulset":
+			c.KubeClientSet.CoreV1().(*fake.FakeCoreV1).PrependReactor("create", "statefulsets",
+				func(action testing2.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &appsv1.StatefulSet{}, errors.New("error creating statefulsets")
+				})
+		}
+
+		err := c.createOrUpdateAffinityAssistantsAndPVCs(ctx, testPRWithVolumeClaimTemplate, tc.aaBehavior)
+
+		if err == nil {
+			t.Errorf("expect error from createOrUpdateAffinityAssistantsAndPVCs but got nil")
+		}
+
+		switch tc.failureType {
+		case "pvc":
+			if !errors.Is(err, ErrPvcCreationFailed) {
+				t.Errorf("expected err type mismatching, expecting %v but got: %v", ErrPvcCreationFailed, err)
+			}
+		case "statefulset":
+			if !errors.Is(err, ErrAffinityAssistantCreationFailed) {
+				t.Errorf("expected err type mismatching, expecting %v but got: %v", ErrAffinityAssistantCreationFailed, err)
+			}
+		}
+		if d := cmp.Diff(tc.expectedErr.Error(), err.Error()); d != "" {
```

Comment:
nit: I think you can compare expectedErr and err directly

Reply:
Here I'm validating the error messages so using `.Error()` (Error type is validated earlier).

It seems like I cannot compare error directly due to [`panic: cannot handle unexported field at {*fmt.wrapError}`](https://github.com/golang/go/blob/ffd9bd7e605cdd2eb66e38bad6e0d93b0d37963c/src/fmt/errors.go#L54C1-L57C2).

Reply:
you can do this with cmpopts.EquateErrors I think? This will check that the error is the right type which is more important than validating the exact error message

Reply:
The type is validate here: https://github.com/tektoncd/pipeline/blob/e0d116852594ed0cf4106da4f2e40fbee2d47b26/pkg/reconciler/pipelinerun/affinity_assistant_test.go#L386C1-L395C4

To be considered equivalent in `cmpopts.EquateErrors()`, they must be the same error object so the validation will fail here.

Reply:
I think cmpopts.EquateErrors uses errors.Is, which is what you're using in the check you linked? In that case you can remove the error string comparison, since I don't know that it adds much value compared to the error type and makes the tests more brittle. This is also fine to leave as is though.